### PR TITLE
Pkce Server Support Detection

### PR DESCRIPTION
**Pull request by yasiruk**

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 1]
```diff
@@ -610,13 +616,43 @@ private String handleOAuthAuthorizationRequest(String clientId, HttpServletReque
...
+            if(pkceChallengeCode != null && !OAuth2Util.validatePKCECodeChallenge(pkceChallengeCode, pkceChallengeMethod)) {
+                return EndpointUtil.getErrorPageURL(OAuth2ErrorCodes.INVALID_REQUEST, "Code challenge used is not up to " +
+                                "RFC 7636 specifications."
+                        , null, null);
+            }
+
+
         }
-        params.setPkceCodeChallenge(pkceChallengeCode);
+                params.setPkceCodeChallenge(pkceChallengeCode);
```
Comment (cdwijayarathna) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: There seems to be a formatting issue here

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java [hunk 2]
```diff
@@ -206,9 +232,14 @@ public OAuthAppDO getAppInformation(String consumerKey) throws InvalidOAuthClien
         PreparedStatement prepStmt = null;
         ResultSet rSet = null;
         OAuthAppDO oauthApp = null;
-
+        boolean isPKCESupportEnabled = OAuth2ServiceComponentHolder.isPkceEnabled();
         try {
-            prepStmt = connection.prepareStatement(SQLQueries.OAuthAppDAOSQLQueries.GET_APP_INFO);
+            if(isPKCESupportEnabled) {
```
Comment (cdwijayarathna) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java]: formatting issues

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java [hunk 3]
```diff
@@ -236,8 +267,10 @@ public OAuthAppDO getAppInformation(String consumerKey) throws InvalidOAuthClien
                     oauthApp.setUser(authenticatedUser);
                     oauthApp.setGrantTypes(rSet.getString(8));
                     oauthApp.setId(rSet.getInt(9));
-                    oauthApp.setPkceMandatory("0".equals(rSet.getString(10)) ? false : true);
-                    oauthApp.setPkceSupportPlain("0".equals(rSet.getString(11)) ? false : true);
+                    if(isPKCESupportEnabled) {
```
Comment (cdwijayarathna) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java]: formatting issues

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java [hunk 4]
```diff
@@ -264,10 +297,15 @@ public OAuthAppDO getAppInformationByAppName(String appName) throws InvalidOAuth
         PreparedStatement prepStmt = null;
         ResultSet rSet = null;
         OAuthAppDO oauthApp = null;
-
+        boolean isPKCESupportEnabled = OAuth2ServiceComponentHolder.isPkceEnabled();
         try {
             int tenantID = CarbonContext.getThreadLocalCarbonContext().getTenantId();
-            prepStmt = connection.prepareStatement(SQLQueries.OAuthAppDAOSQLQueries.GET_APP_INFO_BY_APP_NAME);
+            if(isPKCESupportEnabled) {
```
Comment (cdwijayarathna) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java]: formatting issues

**Comment by cdwijayarathna** [comment]
There are places with no space between if and (. Please fix such places

File: components/org.wso2.carbon.identity.oauth.ui/src/main/java/org/wso2/carbon/identity/oauth/ui/client/OAuthAdminClient.java [hunk 5]
```diff
@@ -98,6 +98,9 @@ public OAuthRevocationResponseDTO revokeAuthzForAppsByRessourceOwner(OAuthRevoca
         return stub.revokeAuthzForAppsByResoureOwner(reqDTO);
     }
 
+    public boolean isPKCESupportedEnabled() throws Exception{
```
Comment (Pushpalanka) [inline components/org.wso2.carbon.identity.oauth.ui/src/main/java/org/wso2/carbon/identity/oauth/ui/client/OAuthAdminClient.java]: formatting issue

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/SQLQueries.java [hunk 6]
```diff
@@ -42,15 +45,28 @@ private SQLQueries(){
 
         public static final String GET_APPS_OF_USER_WITH_TENANTAWARE_OR_TENANTUNAWARE_USERNAME = "SELECT " +
                 "CONSUMER_KEY, CONSUMER_SECRET, APP_NAME, OAUTH_VERSION, CALLBACK_URL, GRANT_TYPES, ID, USERNAME, TENANT_ID, " +
+                "USER_DOMAIN FROM IDN_OAUTH_CONSUMER_APPS WHERE (USERNAME=? OR USERNAME=?) AND TENANT_ID=?";
```
Comment (Pushpalanka) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/SQLQueries.java]: Is there any time we save user name with tenant domain in this table?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponent.java [hunk 7]
```diff
@@ -90,6 +98,17 @@ protected void activate(ComponentContext context) {
         } else {
             log.error("OAuth - ApplicationMgtListener could not be registered.");
         }
+        if(checkPKCESupport()) {
+            OAuth2ServiceComponentHolder.setPkceEnabled(true);
+            if(log.isInfoEnabled()) {
```
Comment (Pushpalanka) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponent.java]: No need to check for info enabled

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/SQLQueries.java (hunk 6 above)
Comment (yasiruk) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/SQLQueries.java]: Yes, we do that when we save an oauth application in `org.wso2.carbon.identity.oauth.da.OAuthAppDAO`

`prepStmt.setString(3, consumerAppDO.getUser().getUserName());
                    prepStmt.setInt(4, IdentityTenantUtil.getTenantId(consumerAppDO.getUser().getTenantDomain()));`

**Pull request merged**